src/components/structures/RoomDirectory.js

```diff
@@ -68,54 +80,81 @@ module.exports = React.createClass({
                     publicRooms: data.chunk,
                     loading: false,
                 });
-                self.forceUpdate();
             }
         });
     },
 
-    componentWillUnmount: function() {
-        // dis.dispatch({
-        //     action: 'ui_opacity',
-        //     sideOpacity: 1.0,
-        //     middleOpacity: 1.0,
-        // });
-    },
+    /**
+     * A limited interface for removing rooms from the directory.
+     * Will set the room to not be publicly visible and delete the
+     * default alias. In the long term, it would be better to allow
+     * HS admins to do this through the RoomSettings interface, but
+     * this needs SPEC-417.
+     */
+    removeFromDirectory: function(room) {
+        var alias = get_display_alias_for_room(room);
+        var name = room.name || alias || "Unnamed room";
 
-    showRoom: function(roomId, roomAlias) {
-        // extract the metadata from the publicRooms structure to pass
-        // as out-of-band data to view_room, because we get information
-        // here that we can't get other than by joining the room in some
-        // cases.
-        var room;
-        if (roomId) {
-            for (var i = 0; i < this.state.publicRooms.length; ++i) {
-                if (this.state.publicRooms[i].room_id == roomId) {
-                    room = this.state.publicRooms[i];
-                    break;
-                }
+        var QuestionDialog = sdk.getComponent("dialogs.QuestionDialog");
+        var ErrorDialog = sdk.getComponent("dialogs.ErrorDialog");
+
+        var room_alias = get_display_alias_for_room(room);
+
+        Modal.createDialog(QuestionDialog, {
+            title: "Remove from Directory",
+            description: `Delete the room alias '${room_alias}' and remove '${name}' from the directory?`,
+            onFinished: (should_delete) => {
+                if (!should_delete) return;
+
+                var Loader = sdk.getComponent("elements.Spinner");
+                var modal = Modal.createDialog(Loader);
+                var step = `remove '${name}' from the directory.`;
+
+                MatrixClientPeg.get().setRoomDirectoryVisibility(room.room_id, 'private').then(() => {
```

Comment:
Not *really*. It would be better to have separate controls to delete the individual aliases for a room, but wanted to get something in as a starter so we can moderate.

There are 2 problems here, basically:

1) Make it possible to remove entries from the public room directory easily, for moderation.
2) Make it possible to recover a room alias if you're a server admin or created the alias.

This is sort of a compromise that solves 1 and solves some cases of 2, but not all. We could either just solve 1 and not delete any of the aliases (although the room admins would easily be able to flip the switch back again) or try to solve some more cases of 2 by deleting all the aliases. It still wouldn't let you recover a room alias if everyone had left the room (as these don't appear on the public directory).

wdyt?